src/main/java/org/terasology/lost/OnSpawnSystem.java

```diff
@@ -15,26 +15,139 @@
+        progressTrackingComponent.addLevel("Tropical seasonal forest", "Lost:well");
+        progressTrackingComponent.addLevel("Lakeshore", "Lost:well");
+        progressTrackingComponent.addLevel("Coast", "Lost:well");
+        progressTrackingComponent.addLevel("Ice", "Lost:well");
+        player.addComponent(progressTrackingComponent);
+    }
+
+    @ReceiveEvent
+    public void onBiomeChange(OnBiomeChangedEvent event, EntityRef player) {
+        console.addMessage("OLD:" + event.getOldBiome().getDisplayName());
+        console.addMessage("NEW:" + event.getNewBiome().getDisplayName());
+        LocationComponent loc = player.getComponent(LocationComponent.class);
+        Vector3f pos = loc.getWorldPosition();
+        ProgressTrackingComponent progressTrackingComponent = player.getComponent(ProgressTrackingComponent.class);
+        int searchRadius = 0;
+        Vector3i ext = new Vector3i(searchRadius, 0, searchRadius);
+        Vector3i desiredPos = new Vector3i(pos.getX(), 1, pos.getZ());
+
+        // try and find somewhere in this region a spot to land
+        Region3i spawnArea = Region3i.createFromCenterExtents(desiredPos, ext);
+        Region worldRegion = LostWorldGenerator.world.getWorldData(spawnArea);
+
+        GraphFacet graphs = worldRegion.getFacet(GraphFacet.class);
+        WhittakerBiomeModelFacet model = worldRegion.getFacet(WhittakerBiomeModelFacet.class);
+        Vector2f pos2d = new Vector2f(pos.getX(), pos.getZ());
+        double d = 1000000;
+        ImmutableVector2f center = null;
+        String biomeName = null;
+        for (Graph g : graphs.getAllGraphs()) {
+            BiomeModel biomeModel = model.get(g);
+            for (org.terasology.polyworld.graph.Region r : g.getRegions()) {
+                WhittakerBiome biome = biomeModel.getBiome(r);
+                if (biome.getDisplayName().contains(event.getNewBiome().getDisplayName())) {
+                    double temp = Math.pow(pos2d.x - pos.getX(), 2) + Math.pow(pos2d.y - pos.getZ(), 2);
+                    if (temp < d) {
+                        d = temp;
+                        center = r.getCenter();
+                        biomeName = biome.getDisplayName();
+                    }
+                }
+
+            }
+        }
+        console.addMessage(center.toString());
+        if ((!progressTrackingComponent.isWellFound())&&progressTrackingComponent.getLevelPrefab(biomeName).contains("well")) {
+            progressTrackingComponent.foundWell = true;
+        }
+
+        if (progressTrackingComponent.isWellFound()) {
+            Prefab prefab =
+                    assetManager.getAsset(progressTrackingComponent.getLevelPrefab(biomeName), Prefab.class).orElse(null);
+            EntityBuilder entityBuilder = entityManager.newBuilder(prefab);
+            EntityRef item = entityBuilder.build();
+            int tempy = (int) Math.ceil(pos.y) + 20;
+            while (true) {
```

Comment:
A simple `while(true)` can be somewhat scary since they're a risk for infinite loops due to bugs or unexpected world features. You might want to plug in some sort of threshold / timeout so there's a guaranteed exit with a graceful error

The if statement below here is also relatively gigantic :-) You might want to see if you can either break it up somewhat or at least add a newline or two